File: lib/core/utils/suno_lyric_phonetic_sanitize.dart
/// Post-process Suno output — phonetic apostrophes + instrumental staging leaks.

import '../constants/audio_environment_data.dart';
import 'suno_lyrics_audio_normalizer.dart';

const _verbPrefixes = [
  'features ',
  'feature ',
  'featured ',
  'playing a ',
  'playing ',
  'play a ',
  'play ',
  'letting the ',
  'let the ',
  'let ',
  'adding ',
  'add ',
  'introducing ',
  'introduce ',
];

String _stripVerbLead(String token) {
  var t = token.trim();
  while (t.isNotEmpty) {
    final lower = t.toLowerCase();
    var stripped = false;
    for (final prefix in _verbPrefixes) {
      if (lower.startsWith(prefix)) {
        t = t.substring(prefix.length).trimLeft();
        stripped = true;
        break;
      }
    }
    if (!stripped) break;
  }
  return t;
}

String _cleanInstrumentalBody(String body) {
  return body
      .split(',')
      .map(_stripVerbLead)
      .where((p) => p.isNotEmpty)
      .join(', ');
}

final _instrumentalBracket = RegExp(
  r'\[(Instrumental(?:\s+Break)?|Instrumental Interlude|Log Drum Break):\s*([^\]]+)\]',
  caseSensitive: false,
);

/// Strip verb-phrase leaks from instrumental-only staging brackets.
String sanitizeInstrumentalStagingTags(String text) {
  return text.replaceAllMapped(_instrumentalBracket, (m) {
    final label = m.group(1)!;
    final body = _cleanInstrumentalBody(m.group(2)!);
    return body.isEmpty ? '[$label]' : '[$label: $body]';
  });
}

final _gospelLane = RegExp(
  r'gospel|worship|praise(?:\s*(?:&|and)\s*worship)?|ccm|christian|southern\s+gospel|afro[- ]?gospel|afro[- ]?praise|contemporary\s+christian',
  caseSensitive: false,
);

final _gospelStagingReplacements = <(RegExp, String)>[
  (RegExp(r'\bsidechain\s+pump\b', caseSensitive: false),
      'Analog VCA glue, warm dynamic leveling'),
  (RegExp(r'\bheavier\s+sidechain\b', caseSensitive: false),
      'heavier sustained strings, warm dynamic leveling'),
  (RegExp(r'\bsidechain\s+slam\b', caseSensitive: false),
      'warm dynamic leveling, analog console glue'),
  (RegExp(r'\bsidechain\b', caseSensitive: false), 'warm dynamic leveling'),
  (RegExp(r'\bdj\s+intro\b', caseSensitive: false),
      'Dead-room isolation, pristine studio environment, close-mic vocal tracking'),
  (RegExp(r'\bdj\s+outro\b', caseSensitive: false),
      'Sustained studio band resolution, clean multi-track fade, trailing organ decay'),
  (RegExp(r'\bmix-out\s+groove\b', caseSensitive: false),
      'Sustained studio band resolution, clean multi-track fade, trailing organ decay'),
  (RegExp(r'\bfull\s+satb\s+choir\s+stack\b', caseSensitive: false),
      'Isolated multi-tracked vocal doubles, Tight double-tracked vocal stacks'),
  (RegExp(r'\bauthentic\s+congregational\s+width\b', caseSensitive: false),
      'Multi-tracked vocal overlays, Isolated multi-tracked vocal doubles'),
  (RegExp(r'\bcongregational\s+trailing\s+ad-libs\b', caseSensitive: false),
      'clean multi-track fade'),
  (RegExp(r'\blive\s+band\s+count-in\b', caseSensitive: false),
      'Dead-room isolation, close-mic vocal tracking'),
  (RegExp(r'\bfilter\s+sweep\b', caseSensitive: false), 'natural room decay'),
  (RegExp(r'\blow-pass\s+sweep\b', caseSensitive: false), 'warm dynamic dip'),
  (RegExp(r'\bsupersaw\b', caseSensitive: false), 'lush sustained live strings'),
];

bool _isGospelLane(String primary, String fusion) {
  final blob = '${primary.trim()} ${fusion.trim()}'.trim();
  if (blob.isEmpty) return false;
  return _gospelLane.hasMatch(blob);
}

String _sanitizeGospelBracketInner(String inner) {
  var out = inner;
  for (final (pattern, replacement) in _gospelStagingReplacements) {
    out = out.replaceAll(pattern, replacement);
  }
  return out;
}

final _sectionMark = RegExp(
  r'^\[(Intro|Verse\s*\d+|Chorus|Final\s+Chorus|Bridge|Outro|Pre-Chorus|Drop|Build)\]\s*$',
  caseSensitive: false,
);
final _sectionHeader = RegExp(r'^\[[^\]]+\]\s*$', multiLine: true);

final _studioVocalReplacements = <(RegExp, String)>[
  (RegExp(r'\bstudio\s+harmonic\s+overlays\b', caseSensitive: false),
      'Wider multi-tracked studio harmonies'),
  (RegExp(r'\bstudio\s+harmonic\s+backing\b', caseSensitive: false),
      'Isolated multi-tracked vocal doubles'),
  (RegExp(r'\bharmonic\s+overlays\b', caseSensitive: false),
      'Wider multi-tracked studio harmonies'),
  (RegExp(r'\bharmonic\s+backing\b', caseSensitive: false),
      'Isolated multi-tracked vocal doubles'),
];

final _introCrowdBans = <(RegExp, String)>[
  (RegExp(r'\bcongregational\b', caseSensitive: false),
      'Isolated multi-tracked vocal doubles'),
  (RegExp(r'\bsanctuary\b', caseSensitive: false), 'Pristine studio environment'),
  (RegExp(r'\bcommunal\b', caseSensitive: false), 'Multi-tracked vocal overlays'),
  (RegExp(r'\bchurch\b', caseSensitive: false), 'Warm studio room'),
  (RegExp(r'\bchoir\b', caseSensitive: false), 'Isolated multi-tracked vocal doubles'),
  (RegExp(r'\bcongregation\b', caseSensitive: false),
      'Tight double-tracked vocal stacks'),
  (RegExp(r'\bsatb\b', caseSensitive: false), 'Isolated multi-tracked vocal doubles'),
  (RegExp(r'\blive\b', caseSensitive: false), 'Studio'),
];

final _tapeNoiseBans = <(RegExp, String)>[
  (RegExp(r'\bsubtle\s+tape\s+hiss\b', caseSensitive: false), 'focused studio room'),
  (RegExp(r'\btape\s+hiss\b', caseSensitive: false), 'dry acoustic room'),
  (RegExp(r'\bvinyl\s+crackle\b', caseSensitive: false),
      'sparse fingerpicked acoustic guitar'),
  (RegExp(r'\broom\s+noise\b', caseSensitive: false), 'dead-room silence'),
];

bool _isStagingBracket(String header) {
  if (header.contains(':')) return false;
  if (RegExp(
    r'^(Intro|Verse\s*\d+|Chorus|Final\s+Chorus|Bridge|Outro|Pre-Chorus|Drop|Build)\s*$',
    caseSensitive: false,
  ).hasMatch(header)) {
    return false;
  }
  return header.contains(',');
}

String _sanitizeStudioBracketInner(
  String inner, {
  required bool gospel,
  required bool early,
  required bool outro,
}) {
  var out = gospel ? _sanitizeGospelBracketInner(inner) : inner;
  for (final (pattern, replacement) in _studioVocalReplacements) {
    out = out.replaceAll(pattern, replacement);
  }
  if (early) {
    for (final (pattern, replacement) in _introCrowdBans) {
      out = out.replaceAll(pattern, replacement);
    }
  }
  if (early || outro) {
    for (final (pattern, replacement) in _tapeNoiseBans) {
      out = out.replaceAll(pattern, replacement);
    }
  }
  return out;
}

/// Strip crowd, harmonic-backing, and tape-noise triggers in studio-isolation mode.
String sanitizeStudioIsolationTags(
  String text, {
  String primaryGenre = '',
  String subGenreFusion = '',
  String audioEnvironmentModeId = '',
}) {
  if (AudioEnvironmentData.isLivePerformance(audioEnvironmentModeId)) return text;
  if (text.isEmpty) return text;

  final gospel = _isGospelLane(primaryGenre, subGenreFusion);
  final lines = text.split('\n');
  final outLines = <String>[];
  var currentSection = '';

  for (final line in lines) {
    final stripped = line.trim();
    final sectionMatch = _sectionMark.firstMatch(stripped);
    if (sectionMatch != null) {
      currentSection = sectionMatch.group(1)!;
      outLines.add(line);
      continue;
    }
    if (stripped.startsWith('[') && _sectionHeader.hasMatch(stripped)) {
      final header = stripped.substring(1, stripped.lastIndexOf(']'));
      if (_isStagingBracket(header)) {
        final early = RegExp(r'^(Intro|Verse\s*1)\s*$', caseSensitive: false)
            .hasMatch(currentSection);
        final outro =
            RegExp(r'^Outro\s*$', caseSensitive: false).hasMatch(currentSection);
        outLines.add(
          '[${_sanitizeStudioBracketInner(header, gospel: gospel, early: early, outro: outro)}]',
        );
        continue;
      }
    }
    outLines.add(line);
  }

  return outLines.join('\n');
}

final _electronicDominant = RegExp(
  r'techno|edm|house|trance|amapiano|trap|dnb|drum\s+and\s+bass|hardstyle|dubstep|electro|future\s+bass|melodic\s+techno',
  caseSensitive: false,
);

final _liveAcoustic = RegExp(
  r'gospel|worship|praise|folk|country|acoustic|americana|bluegrass|singer-songwriter|ccm|christian|southern\s+gospel',
  caseSensitive: false,
);

const _electronicLeakPhrases = [
  'sidechain pump',
  'sidechain slam',
  'sidechain',
  'drum loop',
  '16-bar dj',
  'dj intro',
  'dj outro',
  'mix-out groove',
  'mix out groove',
  'filter sweep',
  'low-pass sweep',
  'supersaw',
  'sub bloom',
  'hat rolls',
  '808 bloom',
  'riser',
];

final _acousticMarker = RegExp(
  r'acoustic|live drum|hammond|choir|satb|organic|folk|strings|piano|guitar|congregational',
  caseSensitive: false,
);

bool _fusionActive(String fusion) {
  final f = fusion.trim().toLowerCase();
  if (f.isEmpty) return false;
  return f != 'none' && f != 'n/a' && f != 'na' && f != '-' && f != '—';
}

bool _hybridSplitDnaActive(String primary, String fusion) {
  return _fusionActive(fusion) && _electronicDominant.hasMatch(primary);
}

bool _strictReconciliationLane(String primary, String fusion) {
  if (_hybridSplitDnaActive(primary, fusion)) return false;
  final blob = '${primary.trim()} ${fusion.trim()}'.trim();
  if (blob.isEmpty) return false;
  return _liveAcoustic.hasMatch(blob) || _isGospelLane(primary, fusion);
}

bool _tokenHasElectronicLeak(String token) {
  final low = token.toLowerCase();
  return _electronicLeakPhrases.any(low.contains);
}

String _reconcileBracketInner(
  String inner, {
  required bool strict,
  required bool hybrid,
  required bool gospel,
}) {
  final parts = inner.split(',').map((p) => p.trim()).where((p) => p.isNotEmpty);
  if (strict || gospel) {
    final innerWork = gospel ? _sanitizeGospelBracketInner(inner) : inner;
    final kept = innerWork
        .split(',')
        .map((p) => p.trim())
        .where((p) => p.isNotEmpty)
        .where((p) => !_tokenHasElectronicLeak(p));
    return kept.join(', ');
  }
  if (hybrid && _acousticMarker.hasMatch(inner)) {
    final kept = parts.where((p) => !_tokenHasElectronicLeak(p));
    return kept.join(', ');
  }
  return inner;
}

/// SUNO V4 §3 Critical Reconciliation — strip EDM leaks from live/gospel staging.
String applyCriticalReconciliation(
  String text, {
  String primaryGenre = '',
  String subGenreFusion = '',
}) {
  if (text.isEmpty) return text;
  final strict = _strictReconciliationLane(primaryGenre, subGenreFusion);
  final hybrid = _hybridSplitDnaActive(primaryGenre, subGenreFusion);
  final gospel = _isGospelLane(primaryGenre, subGenreFusion);
  if (!strict && !hybrid) return text;

  return text.replaceAllMapped(RegExp(r'\[([^\]]+)\]'), (m) {
    final inner = _reconcileBracketInner(
      m.group(1)!,
      strict: strict,
      hybrid: hybrid,
      gospel: gospel,
    );
    return inner.isEmpty ? '[]' : '[$inner]';
  });
}

/// Swap club/EDM staging leaks for sanctuary tokens when genre is gospel/worship.
String sanitizeGospelStagingTags(
  String text, {
  String primaryGenre = '',
  String subGenreFusion = '',
}) {
  return applyCriticalReconciliation(
    text,
    primaryGenre: primaryGenre,
    subGenreFusion: subGenreFusion,
  );
}

/// Full V2 post-output hygiene (phonetics + instrumental + reconciliation).
String sanitizeSunoPostOutput(
  String text, {
  String primaryGenre = '',
  String subGenreFusion = '',
  String audioEnvironmentModeId = '',
}) {
  final phonetic = sanitizeSunoLyricPhonetics(text);
  final instrumental = sanitizeInstrumentalStagingTags(phonetic);
  final isolated = sanitizeStudioIsolationTags(
    instrumental,
    primaryGenre: primaryGenre,
    subGenreFusion: subGenreFusion,
    audioEnvironmentModeId: audioEnvironmentModeId,
  );
  final reconciled = applyCriticalReconciliation(
    isolated,
    primaryGenre: primaryGenre,
    subGenreFusion: subGenreFusion,
  );
  return applyAudioEngineNormalizationToSunoOutput(reconciled);
}

/// Strip trailing apostrophes that glitch Suno TTS.
String sanitizeSunoLyricPhonetics(String text) {
  if (text.isEmpty) return text;

  var out = text;
  const leading = <String, String>{
    "'round": 'around',
    "'bout": 'bout',
    "'cause": 'cause',
    "'em": 'em',
    "'til": 'til',
    '\u2019round': 'around',
    '\u2019bout': 'bout',
    '\u2019cause': 'cause',
    '\u2019em': 'em',
    '\u2019til': 'til',
  };
  for (final e in leading.entries) {
    out = out.replaceAll(e.key, e.value);
  }

  out = out.replaceAllMapped(
    RegExp(r"\b([A-Za-z]+)['\u2018\u2019](?![A-Za-z])"),
    (m) => m.group(1)!,
  );
  return out;
}

File: server/app/suno_lyric_phonetic_sanitize.py
"""Post-process Suno output — phonetic apostrophes + instrumental staging leaks."""

from __future__ import annotations

import re

from app.suno_lyrics_audio_normalizer import apply_audio_engine_normalization_to_suno_output

_APOSTROPHE = "''\u2018\u2019"  # straight + curly open/close quotes

_LEADING_CONTRACTIONS: tuple[tuple[str, str], ...] = (
    ("'round", "around"),
    ("'bout", "bout"),
    ("'cause", "cause"),
    ("'em", "em"),
    ("'til", "til"),
)

# Word letter(s) + apostrophe (not followed by more word chars = not internal like o'clock)
_TRAILING_WORD_APOSTROPHE = re.compile(
    rf"\b([A-Za-z]+)[{re.escape(_APOSTROPHE)}](?![A-Za-z])"
)


def sanitize_suno_lyric_phonetics(text: str) -> str:
    """Remove Suno-glitch apostrophes; expand leading 'round → around."""
    if not text:
        return text

    out = text
    for old, new in _LEADING_CONTRACTIONS:
        out = out.replace(old, new)
        curly = old.replace("'", "\u2019")
        out = out.replace(curly, new)

    out = _TRAILING_WORD_APOSTROPHE.sub(r"\1", out)
    return out


_INSTRUMENTAL_BRACKET = re.compile(
    r"\[(Instrumental(?:\s+Break)?|Instrumental Interlude|Log Drum Break):\s*([^\]]+)\]",
    re.IGNORECASE,
)

_VERB_PREFIXES: tuple[str, ...] = (
    "features ",
    "feature ",
    "featured ",
    "playing a ",
    "playing ",
    "play a ",
    "play ",
    "letting the ",
    "let the ",
    "let ",
    "adding ",
    "add ",
    "introducing ",
    "introduce ",
)


def _strip_verb_lead(token: str) -> str:
    t = token.strip()
    while t:
        lower = t.lower()
        stripped = False
        for prefix in _VERB_PREFIXES:
            if lower.startswith(prefix):
                t = t[len(prefix) :].lstrip()
                stripped = True
                break
        if not stripped:
            break
    return t


def _clean_instrumental_body(body: str) -> str:
    parts = [_strip_verb_lead(part) for part in body.split(",")]
    return ", ".join(part for part in parts if part)


def sanitize_instrumental_staging_tags(text: str) -> str:
    """Strip verb-phrase leaks from instrumental-only staging brackets."""

    def repl(match: re.Match[str]) -> str:
        label = match.group(1)
        body = _clean_instrumental_body(match.group(2))
        return f"[{label}: {body}]" if body else f"[{label}]"

    return _INSTRUMENTAL_BRACKET.sub(repl, text)


_GOSPEL_LANE = re.compile(
    r"gospel|worship|praise(?:\s*(?:&|and)\s*worship)?|ccm|christian|southern\s+gospel|afro[- ]?gospel|afro[- ]?praise|contemporary\s+christian",
    re.IGNORECASE,
)

_GOSPEL_STAGING_REPLACEMENTS: tuple[tuple[re.Pattern[str], str], ...] = (
    (re.compile(r"\bsidechain\s+pump\b", re.I), "Analog VCA glue, warm dynamic leveling"),
    (re.compile(r"\bheavier\s+sidechain\b", re.I), "heavier sustained strings, warm dynamic leveling"),
    (re.compile(r"\bsidechain\s+slam\b", re.I), "warm dynamic leveling, analog console glue"),
    (re.compile(r"\bsidechain\b", re.I), "warm dynamic leveling"),
    (re.compile(r"\bdj\s+intro\b", re.I), "Dead-room isolation, pristine studio environment, close-mic vocal tracking"),
    (re.compile(r"\bdj\s+outro\b", re.I), "Sustained studio band resolution, clean multi-track fade, trailing organ decay"),
    (re.compile(r"\bmix-out\s+groove\b", re.I), "Sustained studio band resolution, clean multi-track fade, trailing organ decay"),
    (re.compile(r"\bfull\s+satb\s+choir\s+stack\b", re.I), "Isolated multi-tracked vocal doubles, Tight double-tracked vocal stacks"),
    (re.compile(r"\bauthentic\s+congregational\s+width\b", re.I), "Multi-tracked vocal overlays, Isolated multi-tracked vocal doubles"),
    (re.compile(r"\bcongregational\s+trailing\s+ad-libs\b", re.I), "clean multi-track fade"),
    (re.compile(r"\blive\s+band\s+count-in\b", re.I), "Dead-room isolation, close-mic vocal tracking"),
    (re.compile(r"\bfilter\s+sweep\b", re.I), "natural room decay"),
    (re.compile(r"\blow-pass\s+sweep\b", re.I), "warm dynamic dip"),
    (re.compile(r"\bsupersaw\b", re.I), "lush sustained live strings"),
)


def _is_gospel_lane(primary: str, fusion: str = "") -> bool:
    blob = f"{primary} {fusion}".strip()
    return bool(blob) and _GOSPEL_LANE.search(blob) is not None


def _sanitize_gospel_bracket_inner(inner: str) -> str:
    out = inner
    for pattern, replacement in _GOSPEL_STAGING_REPLACEMENTS:
        out = pattern.sub(replacement, out)
    return out


_SECTION_MARK = re.compile(
    r"^\[(Intro|Verse\s*\d+|Chorus|Final\s+Chorus|Bridge|Outro|Pre-Chorus|Drop|Build)\]\s*$",
    re.IGNORECASE,
)
_SECTION_HEADER = re.compile(r"^\[[^\]]+\]\s*$", re.MULTILINE)

_STUDIO_VOCAL_REPLACEMENTS: tuple[tuple[re.Pattern[str], str], ...] = (
    (re.compile(r"\bstudio\s+harmonic\s+overlays\b", re.I), "Wider multi-tracked studio harmonies"),
    (re.compile(r"\bstudio\s+harmonic\s+backing\b", re.I), "Isolated multi-tracked vocal doubles"),
    (re.compile(r"\bharmonic\s+overlays\b", re.I), "Wider multi-tracked studio harmonies"),
    (re.compile(r"\bharmonic\s+backing\b", re.I), "Isolated multi-tracked vocal doubles"),
)

_INTRO_CROWD_BANS: tuple[tuple[re.Pattern[str], str], ...] = (
    (re.compile(r"\bcongregational\b", re.I), "Isolated multi-tracked vocal doubles"),
    (re.compile(r"\bsanctuary\b", re.I), "Pristine studio environment"),
    (re.compile(r"\bcommunal\b", re.I), "Multi-tracked vocal overlays"),
    (re.compile(r"\bchurch\b", re.I), "Warm studio room"),
    (re.compile(r"\bchoir\b", re.I), "Isolated multi-tracked vocal doubles"),
    (re.compile(r"\bcongregation\b", re.I), "Tight double-tracked vocal stacks"),
    (re.compile(r"\bsatb\b", re.I), "Isolated multi-tracked vocal doubles"),
    (re.compile(r"\blive\b", re.I), "Studio"),
)

_TAPE_NOISE_BANS: tuple[tuple[re.Pattern[str], str], ...] = (
    (re.compile(r"\bsubtle\s+tape\s+hiss\b", re.I), "focused studio room"),
    (re.compile(r"\btape\s+hiss\b", re.I), "dry acoustic room"),
    (re.compile(r"\bvinyl\s+crackle\b", re.I), "sparse fingerpicked acoustic guitar"),
    (re.compile(r"\broom\s+noise\b", re.I), "dead-room silence"),
)


def _is_staging_bracket(header: str) -> bool:
    if ":" in header:
        return False
    if re.match(r"^(Intro|Verse\s*\d+|Chorus|Final\s+Chorus|Bridge|Outro|Pre-Chorus|Drop|Build)\s*$", header, re.I):
        return False
    return "," in header


def _sanitize_studio_bracket_inner(
    inner: str,
    *,
    gospel: bool,
    early: bool,
    outro: bool,
) -> str:
    out = _sanitize_gospel_bracket_inner(inner) if gospel else inner
    for pattern, replacement in _STUDIO_VOCAL_REPLACEMENTS:
        out = pattern.sub(replacement, out)
    if early:
        for pattern, replacement in _INTRO_CROWD_BANS:
            out = pattern.sub(replacement, out)
    if early or outro:
        for pattern, replacement in _TAPE_NOISE_BANS:
            out = pattern.sub(replacement, out)
    return out


def sanitize_studio_isolation_tags(
    text: str,
    *,
    primary_genre: str = "",
    sub_genre_fusion: str = "",
    audio_environment_mode: str = "",
) -> str:
    """Strip crowd, harmonic-backing, and tape-noise triggers in studio-isolation mode."""
    from app.audio_environment import is_live_performance_mode

    if is_live_performance_mode(audio_environment_mode):
        return text
    if not text:
        return text

    gospel = _is_gospel_lane(primary_genre, sub_genre_fusion)
    lines = text.splitlines()
    out_lines: list[str] = []
    current_section = ""

    for line in lines:
        stripped = line.strip()
        section_match = _SECTION_MARK.match(stripped)
        if section_match:
            current_section = section_match.group(1)
            out_lines.append(line)
            continue
        if stripped.startswith("[") and _SECTION_HEADER.match(stripped):
            header = stripped[1 : stripped.rfind("]")]
            if _is_staging_bracket(header):
                early = bool(
                    re.match(r"^(Intro|Verse\s*1)\s*$", current_section, re.I)
                )
                outro = bool(re.match(r"^Outro\s*$", current_section, re.I))
                cleaned = _sanitize_studio_bracket_inner(
                    header,
                    gospel=gospel,
                    early=early,
                    outro=outro,
                )
                out_lines.append(f"[{cleaned}]")
                continue
        out_lines.append(line)

    return "\n".join(out_lines)


def sanitize_gospel_staging_tags(
    text: str,
    *,
    primary_genre: str = "",
    sub_genre_fusion: str = "",
) -> str:
    """Swap club/EDM staging leaks for sanctuary tokens when genre is gospel/worship."""
    return apply_critical_reconciliation(
        text,
        primary_genre=primary_genre,
        sub_genre_fusion=sub_genre_fusion,
    )


_ELECTRONIC_DOMINANT = re.compile(
    r"techno|edm|house|trance|amapiano|trap|dnb|drum\s+and\s+bass|hardstyle|dubstep|electro|future\s+bass|melodic\s+techno",
    re.IGNORECASE,
)

_LIVE_ACOUSTIC = re.compile(
    r"gospel|worship|praise|folk|country|acoustic|americana|bluegrass|singer-songwriter|ccm|christian|southern\s+gospel",
    re.IGNORECASE,
)

_ELECTRONIC_LEAK_PHRASES: tuple[str, ...] = (
    "sidechain pump",
    "sidechain slam",
    "sidechain",
    "drum loop",
    "16-bar dj",
    "dj intro",
    "dj outro",
    "mix-out groove",
    "mix out groove",
    "filter sweep",
    "low-pass sweep",
    "supersaw",
    "sub bloom",
    "hat rolls",
    "808 bloom",
    "riser",
)

_ACOUSTIC_MARKER = re.compile(
    r"acoustic|live drum|hammond|choir|satb|organic|folk|strings|piano|guitar|congregational",
    re.IGNORECASE,
)


def _fusion_active(fusion: str) -> bool:
    f = fusion.strip().lower()
    if not f:
        return False
    return f not in {"none", "n/a", "na", "-", "—"}


def _hybrid_split_dna_active(primary: str, fusion: str) -> bool:
    return _fusion_active(fusion) and bool(_ELECTRONIC_DOMINANT.search(primary or ""))


def _strict_reconciliation_lane(primary: str, fusion: str) -> bool:
    if _hybrid_split_dna_active(primary, fusion):
        return False
    blob = f"{primary} {fusion}".strip()
    if not blob:
        return False
    return bool(_LIVE_ACOUSTIC.search(blob)) or _is_gospel_lane(primary, fusion)


def _token_has_electronic_leak(token: str) -> bool:
    low = token.lower()
    return any(phrase in low for phrase in _ELECTRONIC_LEAK_PHRASES)


def _reconcile_bracket_inner(
    inner: str,
    *,
    strict: bool,
    hybrid: bool,
    gospel: bool,
) -> str:
    if strict or gospel:
        inner_work = _sanitize_gospel_bracket_inner(inner) if gospel else inner
        parts = [p.strip() for p in inner_work.split(",") if p.strip()]
        parts = [p for p in parts if not _token_has_electronic_leak(p)]
        return ", ".join(parts)
    if hybrid and _ACOUSTIC_MARKER.search(inner):
        parts = [p for p in parts if not _token_has_electronic_leak(p)]
        return ", ".join(parts)
    return inner


def apply_critical_reconciliation(
    text: str,
    *,
    primary_genre: str = "",
    sub_genre_fusion: str = "",
) -> str:
    """SUNO V4 §3 Critical Reconciliation — strip EDM leaks from live/gospel staging brackets."""
    if not text:
        return text
    strict = _strict_reconciliation_lane(primary_genre, sub_genre_fusion)
    hybrid = _hybrid_split_dna_active(primary_genre, sub_genre_fusion)
    gospel = _is_gospel_lane(primary_genre, sub_genre_fusion)
    if not strict and not hybrid:
        return text

    def repl(match: re.Match[str]) -> str:
        inner = _reconcile_bracket_inner(
            match.group(1),
            strict=strict,
            hybrid=hybrid,
            gospel=gospel,
        )
        return f"[{inner}]" if inner else "[]"

    return re.sub(r"\[([^\]]+)\]", repl, text)


def sanitize_suno_post_output(
    text: str,
    *,
    primary_genre: str = "",
    sub_genre_fusion: str = "",
    audio_environment_mode: str = "",
) -> str:
    """Full V2 post-output hygiene (phonetics + instrumental + reconciliation)."""
    text = sanitize_suno_lyric_phonetics(text)
    text = sanitize_instrumental_staging_tags(text)
    text = sanitize_studio_isolation_tags(
        text,
        primary_genre=primary_genre,
        sub_genre_fusion=sub_genre_fusion,
        audio_environment_mode=audio_environment_mode,
    )
    text = apply_critical_reconciliation(
        text,
        primary_genre=primary_genre,
        sub_genre_fusion=sub_genre_fusion,
    )
    return apply_audio_engine_normalization_to_suno_output(text)
